molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java

```diff
@@ -289,59 +326,76 @@ private void compareTargetMetaDatas(EntityType targetRepositoryMetaData, EntityT
 	}
 
 	private void applyMappingsToRepositories(MappingTarget mappingTarget, Repository<Entity> targetRepo,
-			boolean addSourceAttribute)
+			Progress progress)
 	{
 		for (EntityMapping sourceMapping : mappingTarget.getEntityMappings())
 		{
-			applyMappingToRepo(sourceMapping, targetRepo, addSourceAttribute);
+			applyMappingToRepo(sourceMapping, targetRepo, progress);
 		}
 	}
 
-	private void applyMappingToRepo(EntityMapping sourceMapping, Repository<Entity> targetRepo,
-			boolean addSourceAttribute)
+	private void applyMappingToRepo(EntityMapping sourceMapping, Repository<Entity> targetRepo, Progress progress)
 	{
 		EntityType targetMetaData = targetRepo.getEntityType();
 		Repository<Entity> sourceRepo = dataService.getRepository(sourceMapping.getName());
 
+		progress.status(format("Mapping source %s", sourceRepo.getEntityType().getLabel()));
+
 		if (targetRepo.count() == 0)
 		{
-			sourceRepo.forEachBatched(entities -> targetRepo.add(entities.stream()
-					.map(sourceEntity -> applyMappingToEntity(sourceMapping, sourceEntity, targetMetaData,
-							sourceMapping.getSourceEntityType(), addSourceAttribute))), 1000);
+			sourceRepo.forEachBatched(entities ->
+			{
+				mapAndAddEntities(sourceMapping, targetRepo, targetMetaData, entities);
+				progress.increment(1);
+			}, MAPPING_BATCH_SIZE);
 		}
 		else
 		{
-			// FIXME adding/updating row-by-row is a performance bottleneck, this code could do streaming upsert
-			sourceRepo.iterator().forEachRemaining(sourceEntity ->
+			sourceRepo.forEachBatched(entities ->
 			{
-				{
-					Entity mappedEntity = applyMappingToEntity(sourceMapping, sourceEntity, targetMetaData,
-							sourceMapping.getSourceEntityType(), addSourceAttribute);
-					if (targetRepo.findOneById(mappedEntity.getIdValue()) == null)
-					{
-						targetRepo.add(mappedEntity);
-					}
-					else
-					{
-						targetRepo.update(mappedEntity);
-					}
-				}
-			});
+				mapAndUpsertEntities(sourceMapping, targetRepo, targetMetaData, entities);
+				progress.increment(1);
+			}, MAPPING_BATCH_SIZE);
 		}
+	}
 
+	private void mapAndUpsertEntities(EntityMapping sourceMapping, Repository<Entity> targetRepo,
+			EntityType targetMetaData, List<Entity> entities)
+	{
+		entities.forEach(sourceEntity ->
+		{
+			// FIXME adding/updating row-by-row is a performance bottleneck, this code could do streaming upsert
```

Comment:
We can simply split them into two batches here and batch insert/update those, no?